lxd/container_lxc.go

```diff
@@ -1879,34 +1878,53 @@ func (c *containerLXC) deviceStart(deviceName string, rawConfig map[string]strin
 		return nil, err
 	}
 
-	// Ensure nic and infiniband devices are attached if running.
-	if isRunning && shared.StringInSlice(rawConfig["type"], []string{"nic", "infiniband"}) {
-		devName := ""
-		for _, dev := range runConfig.NetworkInterfaces[0] {
-			if dev.Key == "link" {
-				devName = dev.Value
-				break
+	// If container is running, live attach device.
+	if isRunning {
+		if shared.StringInSlice(configCopy["type"], []string{"nic", "infiniband"}) {
+			err = c.deviceAttachNIC(configCopy, runConfig)
+			if err != nil {
+				return nil, err
 			}
 		}
 
-		if devName == "" {
-			return nil, fmt.Errorf("Device didn't provide a link property to use")
+		if runConfig != nil {
+			err = c.runPostStartHooks(runConfig.PostStartHooks)
+			if err != nil {
+				return nil, err
+			}
 		}
+	}
 
-		// Load the go-lxc struct.
-		err = c.initLXC(false)
-		if err != nil {
-			return nil, err
-		}
+	return runConfig, nil
+}
 
-		// Add the interface to the container.
-		err = c.c.AttachInterface(devName, configCopy["name"])
-		if err != nil {
-			return nil, fmt.Errorf("Failed to attach interface: %s to %s: %s", devName, configCopy["name"], err)
+// deviceAttachNIC live attaches a NIC device to a container.
+func (c *containerLXC) deviceAttachNIC(configCopy map[string]string, runConfig *device.RunConfig) error {
+	devName := ""
+	for _, dev := range runConfig.NetworkInterface {
+		if dev.Key == "link" {
+			devName = dev.Value
+			break
 		}
 	}
 
-	return runConfig, nil
+	if devName == "" {
+		return fmt.Errorf("Device didn't provide a link property to use")
+	}
+
+	// Load the go-lxc struct.
+	err := c.initLXC(false)
+	if err != nil {
+		return err
+	}
+
+	// Add the interface to the container.
+	err = c.c.AttachInterface(devName, configCopy["name"])
+	if err != nil {
+		return fmt.Errorf("Failed to attach interface: %s to %s: %s", devName, configCopy["name"], err)
+	}
+
+	return nil
```

Comment:
I don't see how any of the code above is related to `Updates references to proxy devices to use device package` was that meant for another commit?

Reply:
@stgraber this was originally in `deviceStart` but I moved it out when adding the proxy related (although not only for proxy) post start hook code to stop this function getting too big. I wanted to separate the nic specific stuff away from the generic `deviceStart` function. This is how it got into that commit as was caused by adding proxy related linking in deviceStart. I can move it into a new commit if you thing its clearer.